ssl/ssl_lib.c

```diff
@@ -3847,6 +3848,55 @@ static int ssl_session_cmp(const SSL_SESSION *a, const SSL_SESSION *b)
     return memcmp(a->session_id, b->session_id, a->session_id_length);
 }
 
+#ifndef OPENSSL_NO_SSLKEYLOG
+/**
+ * @brief Static initialization for a one-time action to initialize the SSL key log.
+ */
+static CRYPTO_ONCE ssl_keylog_once = CRYPTO_ONCE_STATIC_INIT;
+
+/**
+ * @brief Pointer to a read-write lock used to protect access to the key log.
+ */
+static CRYPTO_RWLOCK *keylog_lock = NULL;
+
+/**
+ * @brief Pointer to a BIO structure used for writing the key log information.
+ */
+static BIO *keylog_bio = NULL;
+
+/**
+ * @brief Ref counter tracking the number of keylog_bio users.
+ */
+static unsigned int keylog_count = 0;
+
+/**
+ * @brief Initializes the SSLKEYLOGFILE lock.
+ *
+ * @return 1 on success, 0 on failure.
+ */
+DEFINE_RUN_ONCE_STATIC(ssl_keylog_init)
+{
+    keylog_lock = CRYPTO_THREAD_lock_new();
+    if (keylog_lock == NULL)
+        return 0;
+    return 1;
+}
+
+static void sslkeylogfile_cb(const SSL *ssl, const char *line)
+{
+    if (keylog_lock == NULL)
+        return;
+
+    if (!CRYPTO_THREAD_write_lock(keylog_lock))
+        return;
+    if (keylog_bio != NULL) {
```

Comment:
Perhaps it would be better to simply create the BIO on-the-fly when we need to log something and then immediately close it when we're finished using it. Yes, there is a performance overhead to do that - but this is debugging tool right? Not something you are going to be doing in production. This might significantly simply the creation and management of the global BIO (because there wouldn't be one).

Reply:
If we're going to go to the trouble of opening/writing/closing a bio on every write to a participating SSL, then it seems just as reasonable to register the callback, and use BIO_up_ref.  Both eliminate the need for external refcounting.

Reply:
>  then it seems just as reasonable to register the callback, and use BIO_up_ref. Both eliminate the need for external refcounting.

Errr....don't you then still have the problem of knowing when the ref count is zero to NULL out the global BIO in that case? Whereas with my proposed alternative there is no global BIO and the whole problem goes away.

Reply:
Ok, I've updated the BIO ref counting to use BIO_up_ref, along with a callback to detect the final free at refcnt == 0, which is actually reasonably simpler, as the callback just has to check if oper is BIO_CB_FREE and set the global pointer to NULL if it is.

I've also adjusted the error handling to return an error in the event anything goes wrong when setting up logging.

Still having an hard time seeing the advantage in converting a cb pointer to a flag here though. I don't really see it saving any significant space or compute time.